crates/evm-nonces-recovery/src/lib.rs

```diff
@@ -0,0 +1,208 @@
+                weight.saturating_accrue(recovery_weight);
+            })
+            .count()
+            .try_into()
+            .expect("Accounts count mustn't overflow");
+        weight.saturating_accrue(retrieval_weight);
+        weight.saturating_accrue(R::DbWeight::get().reads_writes(accounts_count, accounts_count));
+
+        info!("{pallet_name}: Migrated");
+
+        weight
+    }
+
+    #[cfg(feature = "try-runtime")]
+    fn pre_upgrade() -> Result<Vec<u8>, &'static str> {
+        let accounts = <Account<R>>::iter().collect();
+        Ok(PreUpgradeState::<R> { accounts }.encode())
+    }
+
+    #[cfg(feature = "try-runtime")]
+    fn post_upgrade(state: Vec<u8>) -> Result<(), &'static str> {
+        let PreUpgradeState::<R> {
+            accounts: prev_accounts,
+        } = Decode::decode(&mut state.as_slice())
+            .map_err(|_err| "Failed pre-upgrade state decoding")?;
+        for (account_id, prev_account) in prev_accounts {
+            let account = <Account<R>>::try_get(account_id)
+                .map_err(|_: ()| "There should be no lost accounts")?;
+            ensure!(
+                account.data == prev_account.data,
+                "Account's data should remain unchanged",
+            );
+            ensure!(
+                account.nonce >= prev_account.nonce,
+                "Account's nonce shouldn't decrease",
+            );
+        }
+
+        let accounts_to_recover = Self::accounts_managed_by_evm()
+            .filter_map(|(account_id, _weight)| account_id)
+            .filter(|account_id| {
+                <Account<R>>::try_get(account_id)
+                    .map(|account| account.nonce.is_zero())
+                    .unwrap_or(true)
+            })
+            .count();
+        ensure!(
+            accounts_to_recover == 0,
+            "There should be no accounts left for recovery",
+        );
+        Ok(())
+    }
+}
+
+impl<R, Precompiles> MigrationBrokenNoncesRecovery<R, Precompiles>
+where
+    R: pallet_evm::Config,
+    R: pallet_evm_system::Config<AccountId = H160>,
+    Precompiles: TypedGet,
+    Precompiles::Type: IntoIterator<Item = H160>,
+{
+    /// Checks the contract account state and recovers it if necessary.
+    /// - If the state is missing, recreates it.
+    /// - If the state has nonce = 0, writes the smallest possible valid nonce.
+    ///
+    /// Precompiled contracts in Ethereum usually have nonce = 0 as well. However, since precompiled
+    /// contracts are typically implemented by hooking calls to specific addresses and adding dummy
+    /// state (to ensure they are callable like regular contracts), there's no need for a non-zero
+    /// nonce unless they explicitly perform state-changing operations like `CREATE`. Therefore,
+    /// precompiled contracts MUST NOT be passed here.
+    fn recover(id: &H160, account: &mut Option<AccountInfoOf<R>>) -> Weight {
+        let Some(account) = account.as_mut() else {
```

Comment:
Note this branching could be unified if we do a `let account = account.get_or_insert_with(Default::default());` (except for the logging there).